src/system/SystemPacketBuffer.cpp

```diff
@@ -56,6 +56,31 @@
 #else
 #define PBUF_STRUCT_DATA_CONTIGUOUS(pbuf) (pbuf)->type_internal & PBUF_TYPE_FLAG_STRUCT_DATA_CONTIGUOUS
 #endif
+
+#if CHIP_SYSTEM_PACKETBUFFER_FROM_LWIP_RAM_HEAP
+
+struct mem
+{
+    mem_size_t next;
+    mem_size_t prev;
+    u8_t used;
+#if MEM_OVERFLOW_CHECK
+    mem_size_t user_size;
+#endif
+};
+
+#if MEM_OVERFLOW_CHECK
+#define MEM_SANITY_OFFSET MEM_SANITY_REGION_BEFORE_ALIGNED
+#define MEM_SANITY_OVERHEAD (MEM_SANITY_REGION_BEFORE_ALIGNED + MEM_SANITY_REGION_AFTER_ALIGNED)
+#else
+#define MEM_SANITY_OFFSET 0
+#define MEM_SANITY_OVERHEAD 0
+#endif
+#define SIZEOF_STRUCT_MEM LWIP_MEM_ALIGN_SIZE(sizeof(struct mem))
+
+extern "C" uint8_t ram_heap[];
```

Comment:
This is not a generic detail, and should not appear in SystemPacketbuffer.cpp.

Please do not change SystemPacketBuffer.cpp. You should be able to get what you need/want out of `CHIP_SYSTEM_CONFIG_PACKETBUFFER_LWIP_PBUF_TYPE` being updated.

Reply:
Memory heap `ram_heap` and struct mem are originally defined from [lwip/src/core/mem.c](https://github.com/lwip-tcpip/lwip/blob/master/src/core/mem.c). And with the following defines to access:

- PBUF_POOL_SIZE = 0
- MEM_LIBC_MALLOC = 0
- MEM_USE_POOLS = 0

Add `#define CHIP_SYSTEM_CONFIG_PACKETBUFFER_LWIP_PBUF_TYPE PBUF_RAM` in our platform header `SystemPlatformConfig.h` will navigate to mem_malloc/mem_free in lwip/src/core/mem.c for memory access.

But the method **AllocSize()** returns the const value **kMaxSizeWithoutReserve**, which is not the actual memory size allocated from lwip/src/core/mem.c and it may make memory access overflow and crash the system. It needs update method **AllocSize()** for this kind of lwip memory management.

@tcarmelveilleux If my explanation above is something wrong, please point it out.

If my explanation is right, how about to update method **AllocSize()** like below:
```c
    uint16_t AllocSize() const
    {
   ...
#elif CHIP_SYSTEM_PACKETBUFFER_FROM_LWIP_RAM_HEAP
        return lwip_mem_size(&(this->next)) - kStructureSize;
#else
    ...
    }
```
- this->next is the field of pbuf, and &(this->next) is the address of pbuf buffer
- lwip_mem_size is the function implemented by platform to calculate memory size.

Reply:
Have a discussion with @bzbarsky-apple :
 - There is no API or struct exposed from [mem.h](https://github.com/lwip-tcpip/lwip/blob/master/src/include/lwip/mem.h) to get the size of allocated memory
 - Calling lwip_mem_size(this) in AllocSize() is a better way and makes scenes to @bzbarsky-apple. 
 - size_t lwip_mem_size(const void *p_mem) will be implemented and mantained in our platform code. The sample code is as below.
   ```c
    size_t lwip_mem_size(const void *p_mem)
    {
      size_t mem_size = 0;

      sys_mutex_lock(&mem_mutex);
      mem_size = p_mem->next - mem_to_ptr(p_mem);
      sys_mutex_unlock(&mem_mutex);

      return mem_size;
    }
    ```

@tcarmelveilleux Do you have any other suggestions/concerns about this?

Reply:
Passing p_mem as a void * and assuming it to be a pbuf struct seems risky. Why not pass it as a pbuf pointer? Also, p_mem->next could be a null pointer, potentially, isn't it?

Reply:
Function `lwip_mem_size` is used to get memory size in method AllocSize() as below. `this` is the pointer of `struct pbuf`.

```c
    uint16_t AllocSize() const
    {
   ...
#elif CHIP_SYSTEM_PACKETBUFFER_FROM_LWIP_RAM_HEAP
        return lwip_mem_size(this) - kStructureSize;
#else
    ...
    }
```

For debug purpose, sample code `lwip_mem_size` can have a `LWIP_ASSERT` within it, like this.
```c
 size_t lwip_mem_size(const void *p_mem)
 {
   size_t mem_size = 0;

   LWIP_ASSERT("p_mem != NULL", (p_mem != NULL));
   ...
 }
```